anacron/runjob.c

```diff
@@ -399,3 +411,134 @@ tend_children(void)
 	j++;
     }
 }
+
+/* Expand env variables in 'source' arg and save to 'result'*/
+void expand_envvar(const char *source, char *result, const int max_size) {
+    const char *envvar_p;
+    int envvar_name_size = 0;
+
+    *result = '\0';
+
+    while (find_envvar(source, &envvar_p, &envvar_name_size)) {
+        char *envvar_name, *envvar_value;
+        int prefix_size;
+        
+        /* Copy content before env var name */
+        prefix_size = envvar_p - source;
+        
+        if (prefix_size > 0) {
+            if ((strlen(result) + prefix_size + 1) > max_size) {
+                goto too_big;
+            }
+            strncat(result, source, prefix_size);
+        }
+        
+        /* skip envvar name */
+        source = envvar_p + envvar_name_size;
+        
+        /* copy envvar name, ignoring $, { and } chars*/
+        envvar_p++; 
+        envvar_name_size--;
+      
+        if (*envvar_p == '{') {
+            envvar_p++;
+            envvar_name_size = envvar_name_size - 2;
+        }
+
+        envvar_name = malloc(sizeof(char) * (envvar_name_size + 1));
+        strncpy(envvar_name, envvar_p, envvar_name_size);
+        envvar_name[envvar_name_size] = '\0';
+
+        /* Copy envvar value to result */
+        envvar_value = getenv(envvar_name);
+        free(envvar_name);
+        
+        if (envvar_value != NULL) {
+            if ((strlen(result) + strlen(envvar_value) + 1) > max_size) {
+                goto too_big;
+            }
+            strncat(result, envvar_value, strlen(envvar_value));
+        }
+    }
+
+    /* Copy any character left in the source string */
+    if (*source != '\0') {
+        if ((strlen(result) + strlen(source) + 1) > max_size) {
+            goto too_big;
+        }
+        strncat(result, source, max_size);
+    }
+    
+    return;
+    too_big:
+        die_e("Environment variable expansion failed. Expanded value is bigger than allocated memory.");
+}
+
+/* Return:
+ * 0 - Not found
+ * 1 - Found */
+int find_envvar(const char *source, const char **start_pos, int *length) {
```

Comment:
This should be a static int function. Please also put it before the expand_envvar and remove the declaration from the header files.